com.unity.render-pipelines.universal/ShaderLibrary/Shadows.hlsl

```diff
@@ -1,321 +1,374 @@
+    attenuation += fetchesWeights[7] * SAMPLE_TEXTURE2D_SHADOW(ShadowMap, sampler_ShadowMap, float3(fetchesUV[7].xy, shadowCoord.z));
+    attenuation += fetchesWeights[8] * SAMPLE_TEXTURE2D_SHADOW(ShadowMap, sampler_ShadowMap, float3(fetchesUV[8].xy, shadowCoord.z));
+#endif
+
+    return attenuation;
+}
+
+real SampleShadowmap(TEXTURE2D_SHADOW_PARAM(ShadowMap, sampler_ShadowMap), float4 shadowCoord, ShadowSamplingData samplingData, half4 shadowParams, bool isPerspectiveProjection = true)
+{
+    // Compiler will optimize this branch away as long as isPerspectiveProjection is known at compile time
+    if (isPerspectiveProjection)
+        shadowCoord.xyz /= shadowCoord.w;
+
+    real attenuation;
+    real shadowStrength = shadowParams.x;
+
+#ifdef _SHADOWS_SOFT
+    if(shadowParams.y != 0)
+    {
+        attenuation = SampleShadowmapFiltered(TEXTURE2D_SHADOW_ARGS(ShadowMap, sampler_ShadowMap), shadowCoord, samplingData);
+    }
+    else
+#endif
+    {
+        // 1-tap hardware comparison
+        attenuation = SAMPLE_TEXTURE2D_SHADOW(ShadowMap, sampler_ShadowMap, shadowCoord.xyz);
+    }
+
+    attenuation = LerpWhiteTo(attenuation, shadowStrength);
+
+    // Shadow coords that fall out of the light frustum volume must always return attenuation 1.0
+    // TODO: We could use branch here to save some perf on some platforms.
+    return BEYOND_SHADOW_FAR(shadowCoord) ? 1.0 : attenuation;
+}
+
+half ComputeCascadeIndex(float3 positionWS)
+{
+    float3 fromCenter0 = positionWS - _CascadeShadowSplitSpheres0.xyz;
+    float3 fromCenter1 = positionWS - _CascadeShadowSplitSpheres1.xyz;
+    float3 fromCenter2 = positionWS - _CascadeShadowSplitSpheres2.xyz;
+    float3 fromCenter3 = positionWS - _CascadeShadowSplitSpheres3.xyz;
+    float4 distances2 = float4(dot(fromCenter0, fromCenter0), dot(fromCenter1, fromCenter1), dot(fromCenter2, fromCenter2), dot(fromCenter3, fromCenter3));
+
+    half4 weights = half4(distances2 < _CascadeShadowSplitSphereRadii);
+    weights.yzw = saturate(weights.yzw - weights.xyz);
+
+    return 4 - dot(weights, half4(4, 3, 2, 1));
+}
+
+float4 TransformWorldToShadowCoord(float3 positionWS)
+{
+#ifdef _MAIN_LIGHT_SHADOWS_CASCADE
+    half cascadeIndex = ComputeCascadeIndex(positionWS);
+#else
+    half cascadeIndex = 0;
+#endif
+
+    return mul(_MainLightWorldToShadow[cascadeIndex], float4(positionWS, 1.0));
+}
+
+half MainLightRealtimeShadow(float4 shadowCoord)
+{
+#if !defined(MAIN_LIGHT_CALCULATE_SHADOWS)
+    return 1.0h;
+#endif
+
+    ShadowSamplingData shadowSamplingData = GetMainLightShadowSamplingData();
+    half4 shadowParams = GetMainLightShadowParams();
+    return SampleShadowmap(TEXTURE2D_ARGS(_MainLightShadowmapTexture, sampler_MainLightShadowmapTexture), shadowCoord, shadowSamplingData, shadowParams, false);
+}
+
+// returns 0.0 if position is in light's shadow
+// returns 1.0 if position is in light
+half AdditionalLightRealtimeShadow(int lightIndex, float3 positionWS, half3 lightDirection)

```

Comment:
This will be a breaking change for anyone currently calling `AdditionalLightRealtimeShadow` by adding the lightDirection.

Reply:
Good catch.
To prevent user shader compilation errors, I added a version of the old API in commit dabaf72

Reply:
Great! You can move the old API to deprecated.hlsl :)

Reply:
I tried to move the old `AdditionalLightRealtimeShadow(int lightIndex, float3 positionWS)` API to Deprecated.hlsl, but since it references the new version of the API, it causes compilation errors in shaders that include Deprecated.hlsl but not Shadows.hlsl.

Do you think Shadows.hlsl should be included into every shader that refers to Deprecated.hlsl? This might make them a bit slower to process at build time ; and on platforms that support it, could make low-level shader debugging a bit more cumbersome.

Reply:
Ahh I see. No let's then keep it the way it is. 
I will take up the discussion of removing deprecated things after the next major bump. And since you commented it as deprecated I will find it once we get to that point :)
All good!